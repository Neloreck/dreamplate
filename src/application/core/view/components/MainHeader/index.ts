/**
 * @packageDocumentation
 * @module @application/main
 */

export * from "./MainHeader.component";
