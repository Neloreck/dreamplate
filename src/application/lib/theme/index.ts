export * from "@/lib/theme/types";
export * from "@/lib/theme/utils";
