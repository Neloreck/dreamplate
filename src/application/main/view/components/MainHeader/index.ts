export { MainHeader, IMainHeaderInjectedProps } from "./MainHeader.Component";
