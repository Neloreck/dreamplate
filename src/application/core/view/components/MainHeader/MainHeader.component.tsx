import { useManager } from "dreamstate";
import { ReactElement } from "react";
import { createUseStyles } from "react-jss";

import { IApplicationTheme } from "@Lib/theme";

import { ThemeManager } from "@Core/data/store";

export const useStyles = createUseStyles(({ palette, spacing }: IApplicationTheme) => ({
  root: {
    "& > header": {
      color: GColor.BLACK,
      cursor: "default"
    },
    background: palette.primary.dark,
    boxShadow: "0px 2px 4px -1px rgba(0, 0, 0, 0.25), 0px 4px 5px 0px rgba(0, 0, 0, 0.15)," +
            "0px 1px 10px 0px rgba(0, 0, 0, 0.10)",
    boxSizing: "border-box",
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    minHeight: spacing.unit * 10,
    padding: `${spacing.unit * 2}px ${spacing.unit * 4}px`,
    position: "relative"
  }
}));

export function MainHeader({
  classes: { root } = useStyles(),
  themeContext: { themeActions: { toggleTheme } } = useManager(ThemeManager)
}): ReactElement {
  return (
    <header className={root}>
      <h2> Dreamplate </h2>

      <button onClick={toggleTheme}> Toggle </button>
    </header>
  );
}
