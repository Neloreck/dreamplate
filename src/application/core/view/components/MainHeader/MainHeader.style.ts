/**
 * @packageDocumentation
 * @module @application/main
 */

import { createUseStyles } from "react-jss";

// Lib.
import { IApplicationTheme } from "@Lib/theme";

export const useStyles = createUseStyles(({ palette, spacing }: IApplicationTheme) => ({
  root: {
    "& > custom-heading": {
      color: GColor.BLACK,
      cursor: "default"
    },
    alignSelf: "normal",
    background: palette.primary.dark,
    boxShadow: "0px 2px 4px -1px rgba(0, 0, 0, 0.25), 0px 4px 5px 0px rgba(0, 0, 0, 0.15), 0px 1px 10px 0px rgba(0, 0, 0, 0.10)",
    boxSizing: "border-box",
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    minHeight: spacing.unit * 15,
    padding: `${spacing.unit * 2}px ${spacing.unit * 4}px`,
    position: "relative",
    transitionDuration: "250ms"
  }
}));
