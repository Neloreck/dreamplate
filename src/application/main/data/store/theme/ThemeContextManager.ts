import { Bind, ContextManager } from "dreamstate";

// Lib.
import { Logger } from "@Lib/utils";

// Data.
import { EThemeType, IApplicationTheme } from "@Main/data/store/theme/ThemeTypes";
import { createTheme } from "@Main/data/store/theme/ThemeUtils";

export interface IThemeContext {
  themeActions: {
    isDark(): boolean;
    toggleTheme(): void;
  };
  themeState: {
    theme: IApplicationTheme;
  };
}

export class ThemeContextManager extends ContextManager<IThemeContext> {

  private static readonly DEFAULT_THEME_PARAMS: IApplicationTheme = {
    palette: {
      background: {
        default: "#FFF",
        paper: "#EEE"
      },
      primary: {
        dark: "#4c497c",
        light: "#9690FC",
        main: "#a8a5ec",
      },
      secondary: {
        dark: "#6b5b71",
        light: "rgb(166,143,176)",
        main: "#cfb3dc",
      },
      text: {
        primary: "#000",
        secondary: "#333"
      },
      type: EThemeType.LIGHT
    },
    spacing: {
      unit: 8
    }
  };

  protected context: IThemeContext = {
    themeActions: {
      isDark: this.isDark,
      toggleTheme: this.toggleTheme
    },
    themeState: {
      theme: createTheme(ThemeContextManager.DEFAULT_THEME_PARAMS)
    }
  };

  private readonly log: Logger = new Logger("[üç¨THEME]");
  private readonly setState = ContextManager.getSetter(this, "themeState");

  protected onProvisionStarted(): void {
    this.log.info("Started theme context provision.");
  }

  @Bind()
  private isDark(): boolean {
    return this.context.themeState.theme.palette.type === "dark";
  }

  @Bind()
  private toggleTheme(): void {

    const { theme } = this.context.themeState;

    const nextThemeType: EThemeType = (theme.palette.type === EThemeType.LIGHT ? EThemeType.DARK : EThemeType.LIGHT);

    theme.palette.type = nextThemeType;

    this.log.info(`Toggle to '${nextThemeType}'.`);

    this.setState({ theme: createTheme(theme) });
  }

}
