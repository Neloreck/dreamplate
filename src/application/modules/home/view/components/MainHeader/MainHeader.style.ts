import {BLACK, IApplicationTheme, WHITE} from "@Lib/theme";

export const mainHeaderStyle = ({ palette, spacing }: IApplicationTheme) => ({
  root: {
    "& > h2": {
      "-webkit-text-stroke": "1px black",
      cursor: "default",
      fontSize: spacing.unit * 12,
      fontWeight: 600
    },
    alignSelf: "normal",
    alignItems: "center",
    background: palette.primary.dark,
    boxShadow: `0px 2px 4px -1px rgba(0, 0, 0, 0.25), 0px 4px 5px 0px rgba(0, 0, 0, 0.15), 0px 1px 10px 0px rgba(0, 0, 0, 0.10)`,
    boxSizing: "border-box",
    color: WHITE,
    display: "flex",
    flexDirection: "row",
    fontSize: spacing.unit * 8,
    justifyContent: "space-between",
    minWidth: spacing.unit * 60,
    padding: `${spacing.unit * 2}px ${spacing.unit * 6}px`,
    position: "relative",
    transitionDuration: "250ms"
  },
  switchButton: {
    "&:hover": {
      filter: "brightness(0.85)"
    },
    border: `1px solid ${BLACK}`,
    color: BLACK,
    background: palette.primary.light,
    boxShadow: `6px 1px 4px 0px rgba(0, 0, 0, 0.25), 6px 7px 5px 0px rgba(0, 0, 0, 0.2)`,
    cursor: "pointer",
    fontSize: spacing.unit * 4,
    fontWeight: "bold",
    padding: spacing.unit * 3,
    transitionDuration: "250ms"
  }
});
