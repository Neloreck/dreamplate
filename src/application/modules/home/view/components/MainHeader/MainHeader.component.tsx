/**
 * @module @application/main
 */

import { PureComponent, ReactNode } from "react";
import { WithStyles } from "react-jss";

// Lib.
import { Styled } from "@Lib/decorators";

// Data.
import { themeContextManager } from "@Main/data/store";

// View.
import { mainHeaderStyle } from "./MainHeader.style";

// Props.
export interface IMainHeaderOwnProps {}

export interface IMainHeaderInjectedProps extends WithStyles<typeof mainHeaderStyle> {}

export interface IMainHeaderProps extends IMainHeaderOwnProps, IMainHeaderInjectedProps {}

@Styled(mainHeaderStyle)
export class MainHeader extends PureComponent<IMainHeaderProps> {

  public render(): ReactNode {

    const { classes } = this.props;

    return (
      <header className={classes.root}>

        <h2>
          Boilerplate
        </h2>

        <button className={classes.switchButton} onClick={themeContextManager.toggleTheme}>
          Toggle
        </button>

      </header>
    );
  }

}
