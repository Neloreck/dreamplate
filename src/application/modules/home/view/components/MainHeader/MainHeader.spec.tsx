// Test.
import { getThemedComponent, withMockedTheme } from "@Lib/theme/testing";
import { shallow } from "enzyme";

// View.
import { IMainHeaderInjectedProps, MainHeader } from "./MainHeader.component";

describe("Main header component rendering.", () => {

  it("Should match HOC snapshot.", () => {

    const wrapper = shallow(<MainHeader {...{} as IMainHeaderInjectedProps}/>);

    expect(wrapper).toMatchSnapshot("Default HOC layout.");
  });

  it("Should match mounted snapshot with default theme.", () => {

    const wrapper = shallow(withMockedTheme(<MainHeader {...{} as IMainHeaderInjectedProps}/>));
    const mounted = getThemedComponent(wrapper);

    expect(mounted).toMatchSnapshot("Mounted with default theme.");
  });
});
