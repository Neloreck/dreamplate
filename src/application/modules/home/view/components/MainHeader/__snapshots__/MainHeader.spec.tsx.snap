// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main header component rendering. Should match HOC snapshot.: Default HOC layout. 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Main header component rendering. Should match mounted snapshot with default theme.: Mounted with default theme. 1`] = `
<MainHeader
  classes={
    Object {
      "root": "MainHeader-root-0-2-1",
    }
  }
>
  <header
    className="MainHeader-root-0-2-1"
  >
    <custom-heading
      size={3}
      text="Boilerplate"
    />
    <custom-button
      onClick={[Function]}
    >
      Toggle
    </custom-button>
  </header>
</MainHeader>
`;
